contracts/KyberPoolMaster.sol

```diff
@@ -298,88 +403,211 @@ contract KyberPoolMaster is Ownable {
+        uint256[] memory result = new uint256[](epochCounter);
+        for (uint256 i = 0; i < epochCounter; i++) {
+            result[i] = epochsWithRewards[i];
+        }
 
-        uint256 totalFee = totalRewards.mul(epochDFee.fee).div(
-            MAX_DELEGATION_FEE
-        );
-        uint256 rewardsAfterFee = totalRewards.sub(totalFee);
+        return result;
+    }
 
-        uint256 poolMembersShare = calculateRewardsShare(
-            delegatedStake,
-            stake.add(delegatedStake),
-            rewardsAfterFee
-        );
-        uint256 poolMasterShare = totalRewards.sub(poolMembersShare); // fee + poolMaster stake share
+    struct RewardInfo {
+        IExtendedKyberFeeHandler kyberFeeHandler;
+        IERC20 rewardToken;
+        uint256 initialBalance;
+        uint256 totalRewards;
+        uint256 totalFee;
+        uint256 rewardsAfterFee;
+        uint256 poolMembersShare;
+        uint256 poolMasterShare;
+    }
 
-        claimedPoolReward[epoch] = true;
-        memberRewards[epoch] = Reward(poolMembersShare, delegatedStake);
+    /**
+     * @dev  Claims rewards for a given group of epochs in all feeHandlers, distribute fees and its share to poolMaster
+     * @param _epochGroup An array of epochs for which rewards are being claimed
+     */
+    function claimRewardsMaster(uint256[] memory _epochGroup) public {
+        require(_epochGroup.length > 0, "cRMaste: _epochGroup required");
 
-        // distribute poolMasterRewards to poolMaster
-        address payable poolMaster = payable(owner());
-        require(
-            poolMaster.send(poolMasterShare),
-            "cRMaste: poolMaster share transfer failed"
+        IERC20[] memory tokensWithRewards = new IERC20[](
+            feeHandlersList.length
         );
-
-        if (!epochDFee.applied) {
-            applyFee(epochDFee);
+        uint256 tokensWithRewardsLength = 0;
+        uint256[] memory accruedByToken = new uint256[](feeHandlersList.length);
+
+        for (uint256 j = 0; j < _epochGroup.length; j++) {
+            uint256 _epoch = _epochGroup[j];
+            DFeeData storage epochDFee = delegationFees[getEpochDFeeDataId(
+                _epoch
+            )];
+
+            (uint256 stake, uint256 delegatedStake, ) = kyberStaking
+                .getStakerRawData(address(this), _epoch);
+
+            RewardInfo memory rewardInfo;
+
+            for (uint256 i = 0; i < feeHandlersList.length; i++) {
+                rewardInfo.kyberFeeHandler = IExtendedKyberFeeHandler(
+                    feeHandlersList[i]
+                );
+                uint256 unclaimed = getUnclaimedRewards(
+                    _epoch,
+                    rewardInfo.kyberFeeHandler
+                );
+
+                if (unclaimed > 0) {
+                    rewardInfo
+                        .rewardToken = rewardTokenByFeeHandle[feeHandlersList[i]];
+                    rewardInfo.initialBalance = rewardInfo.rewardToken ==
```

Comment:
Can we extract this logic of getting the balance to a separate method? Something like:
```solidity
function getBalance(address _token) view returns (uint256) {
  return _token == ETH_TOKEN_ADDRESS
                        ? address(this).balance
                        : _token.balanceOf(address(this));
}
```

So it can be also used in line `498`